packages/text-serializer/lib/serializeText.js

```diff
@@ -0,0 +1,42 @@
+function collectElementByIDs(element) {
+  const dataStructures = {};
+  const { parents } = element;
+
+  if (!parents || parents.isEmpty) {
+    return dataStructures;
+  }
+
+  const rootElement = parents.get(parents.length - 1);
+
+  if (rootElement) {
+    rootElement.recursiveChildren.forEach((element) => {
+      // eslint-disable-next-line no-underscore-dangle
+      const isNotEmptyStringElement = element._meta && element._meta.get('id');
+
+      if (isNotEmptyStringElement) {
+        dataStructures[element.id.toValue()] = element;
+      }
+    });
+  }
+
+  return dataStructures;
+}
+
+const isPrimitive = element => element && element.element.match(/^(string|number|boolean|null)$/);
+
+function serializeText(element, done) {
+  if (element.element === 'dataStructure') {
+    return serializeText(element.content, done);
+  }
+
+  const dataStructures = collectElementByIDs(element);
+  if (isPrimitive(element) || isPrimitive(dataStructures[element.element])) {
```

Comment:
Unfortunately I don't think checking that the element is primitive in this way would catch all cases. It might be better to go back to similar logic to what you had before where you are checking the result of valueOf. Since `valueOf` will handle all  other types of elements such as enum of string, ref element, or multiple levels of inheritance.

I think the following cases would demonstrate:

```js
const element = new Enum();
element.enumerations = ['north', 'east', 'south', 'west'];

element.valueOf(); // north (string)
```

Or more levels of recursion:

```js
const message = new StringElement('hello world');
message.id = 'Message';

const error = new Element();
error.element = 'message';
error.id = 'Error';

const notFoundError = new Element();
error.element = 'error';

notFoundError.valueOf(); // hello world (string)
// resolving notFoundError should resolve as a string from Message
```

There's possibly another case with `ref` element too.

Reply:
Addressed in 2d3de2b